### PR TITLE
[FIX] sale_loyalty: skip paid calc if ewallet type

### PR DESCRIPTION
**Pull request by ethanrobv**
**Current behavior:**
Creating an ewallet loyalty program and removing the default
value for `trigger_product_ids` without adding another in its
place will result in any ewallet created using that program to
not have its balance deducted after being used in a sale order.

**Expected behavior:**
The default product should not be required here, nor should any
product, seeing as there is no `required` constraint on this
field nor any related ones.

**Steps to reproduce:**
1. Create an ewallet loyalty program and remove the Top-up
     ewallet product

2. Create an ewallet, give it some balance, and use it in an
     order

3. Check the balance of the wallet to see the erroneous behavior

**Cause of the issue:**
In the _program_check_compute_points() method of `sale.order` in
`sale_loyalty`, the conditional block:

`if not products_per_rule.get(rule):
    continue`

will normally prevent rules belonging to ewallet program types
from going further in the method because they have the default
ewallet top-up product in their domain. When the continue is not
reached, they will reach this line:

`amount_paid = sum(max(0, line.price_total) for
    line in order_lines if line.product_id in rule_products)`

which will end up offsetting the actual subtraction of a SOLs
`points_cost` from an ewallet `loyalty.card`'s balance.

**Fix:**
Add another check prior to the calculation of order points which
prevents orders with applied ewallet coupons without any top-up
products from reaching the problematic code. If an ewallet
program doesn't have any `trigger_product_ids`, it shouldn't
ever need to perform such calculations.

opw-3756134

### CODE REVIEW

**Review by Demesmaeker — COMMENTED** [review]
Hello,

I don't know if it's realistic, but if only part of the order is paid with an eWallet, it'd be strange to not award the points for the rest of the order :thinking: 

What should be the behavior here @dalaOdoo ?

**Comment by dalaOdoo** [comment]
@Demesmaeker It looks similar to a problem @anko-odoo fixed in this PR https://github.com/odoo/odoo/pull/140112 but only for "gift cards"

When there is no product, the rule to grant point is applied on all product instead of none...

**Comment by ethanrobv** [comment]
> Hello,
> 
> I don't know if it's realistic, but if only part of the order is paid with an eWallet, it'd be strange to not award the points for the rest of the order 🤔
> 
> What should be the behavior here @dalaOdoo ?

I think that's sensible, where do you think those points should be calculated? The way I see it the system is set up to spend earned points in the current order which doesn't really make sense IMO.

And thanks for the feedback :+1:

**Comment by ethanrobv** [comment]
@Demesmaeker @dalaOdoo I was actually conflating loyalty card and ewallet programs when I added the TODO comment. I don't believe ewallets accrue points like this. The bug is actually just caused by not accounting for the fact that an ewallet coupon would have reached that block of code because they ordinarily hit the `continue` due to their product domain.

This fix should be good to go (IMO) :+1:

**Comment by dalaOdoo** [comment]
@ethanrobv I know it's weird but with your fix, if a customer use the rest of his eWallet and top up his eWallet in the same order it does not work.
https://drive.google.com/file/d/1Vhq5cGkLn9ovj0qnk3wicQD7acgZ5lXm/view?usp=drivesdk

**Comment by ethanrobv** [comment]
> I know it's weird but with your fix, if a customer use the rest of his eWallet and top up his eWallet in the same order it does not work.

Thanks for finding this case :+1: I added a test for this case as well. I had to modify an existing test after modifying setUpClass(), I know this is not really advisable but I ensured the test was equivalent with and without this most recent version of my fix applied so I think it is OK. LMK if you would rather me revert these changes and keep the record initialization in the tests added specifically for this PR.

**Comment by ethanrobv** [comment]
@dalaOdoo I believe I am waiting for a review from you, thanks in advance :+1:

**Comment by dalaOdoo** [comment]
LGTM

**Review by Feyensv — APPROVED** [review]
@robodoo delegate+

**Comment by ethanrobv** [comment]
@robodoo r+

**Pull request closed without merging**
